src/hardware/dac.rs

```diff
@@ -75,14 +77,39 @@ static mut DAC_BUF: [[SampleBuffer; 2]; 2] = [[[0; SAMPLE_BUFFER_SIZE]; 2]; 2];
 #[derive(Copy, Clone)]
 pub struct DacCode(pub u16);
 
-#[allow(clippy::from_over_into)]
-impl Into<f32> for DacCode {
-    fn into(self) -> f32 {
+impl TryFrom<f32> for DacCode {
+    type Error = ();
+
+    fn try_from(voltage: f32) -> Result<DacCode, ()> {
+        // The DAC output range in bipolar mode (including the external output op-amp) is +/- 4.096
+        // V with 16-bit resolution. The anti-aliasing filter has an additional gain of 2.5.
+        let dac_range = 4.096 * 2.5;
+
+        if voltage > dac_range {
+            Err(())
+        } else if voltage < -1. * dac_range {
+            Err(())
+        } else {
+            Ok(DacCode::from(
+                (voltage / dac_range * i16::MAX as f32) as i16,
```

Comment:
```suggestion
                (voltage * (i16::MAX as f32 / dac_range)) as i16,
```

Or compute `const dac_lsb_per_volt` like in `From<DacCode> for f32`.